### DISCUSSION TITLE
What's the best way to perform bulk upserts?

### DISCUSSION THREAD

**Discussion by persona94** [sqlalchemy/sqlalchemy#11494] [Usage Questions]
I have a table in the following format

The database is setup in a file called db.py

```
dbase = SQLAlchemy()
```

```
from db import dbase as db

class Test(db.Model):
    __tablename__ = 'test'

    test_id = db.Column(db.String, primary_key=True)
    test_name = db.Column(db.String)
    test_enabled = db.Column(db.Boolean)
    test_result = db.Column(db.String)
```

Database set up in a file called 
I'm reading a stream that has a large list of test objects in json format
```
{"test_id":"aaaa", "test_name":"test_1", "test_result":"running"}
{"test_id":"aabb", "test_name":"test_2", "test_result":"running"}
{"test_id":"aaaa", "test_name":"test_1", "test_result":"complete"}
```

I can have several million json rows so I'm trying to batch insert them into my database

```
def add_test_res(test_list):
    db.session.bulk_insert_mappings(Test, test_list)
    db.session.commit()
```

Since I have so many records in to process I split it up in batches of 30k. Is there any way to replace `db.session.bulk_insert_mappings` with a `bulk upsert`?

**Comment by CaselIT** [comment]
Hi,

Have you looked at the docs? https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html

**Reply by persona94** [reply]
I have. There are 2 issues

1. I don't use statements, relying directly on lists to be sent to the session objects. I could start using statements but I'm not able to understand the next issue which is,
2. What should I pass into on_conflict_do_update in my case? The example you linked and nearly every stackoverflow answer talks about excluding 1 field. My json could have multiple fields that have changed for the same ID

**Reply by CaselIT** [reply]
so upsert capabilities depend on the db in use. In all case they require the use of statements. the session does not have that capability.

For postgresql there is this https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.Insert.excluded that let you target the values that would be inserted for the update

**Comment by gordthompson** [comment]
IMO, "the best way to perform bulk upserts" is to upload the source data to a temporary table, and then run the necessary DML statement(s) on the server, e.g.,

- `INSERT INTO … SELECT … ON CONFLICT` for PostgreSQL,
- `MERGE`  for MS SQL Server,
- etc.

**Comment by pavel-v-sobolev** [comment]
Hi! 

I created a python library on top of sqlalchemy for doing mass upserts in effecient way.
It does bulk insert into temporary table, then performs Update/Insert/Delete(if needed).
It works with different database engines.
I hope, you find it useful for such a task!

github link: [DBMerge](https://github.com/pavel-v-sobolev/dbmerge)

Install:
```
pip install dbmerge
```
How to use:
```
from sqlalchemy import create_engine
from datetime import date
from dbmerge import dbmerge

# 1. Initialize DB engine
engine = create_engine("sqlite://")

# 2. Prepare your source data
data = [
    {'Shop': '123', 'Product': 'A1', 'Date': date(2025, 1, 1), 'Qty': 2, 'Price': 50.10},
    {'Shop': '124', 'Product': 'A1', 'Date': date(2025, 1, 1), 'Qty': 1, 'Price': 100.50}
]

# 3. Execute the merge operation
# The table will be created automatically if it doesn't exist.
with dbmerge(engine=engine, data=data, table_name="Facts", 
             key=['Shop', 'Product', 'Date']) as merge:
    merge.exec()
```
